### ISSUE TITLE
Automatically set GOMAXPROCS and GOMEMLIMIT based on available resources (container cgroups / system)

### ISSUE THREAD

**Issue by frittentheke** [grafana/k6#3941]
### Feature Description

While k6 itself does take the number of CPUs into account, e.g. at https://github.com/grafana/k6/blob/f49504674362639fa312d9b5bd8ecf8d8f93ddbd/execution/scheduler.go#L268 (planned at https://github.com/grafana/k6/blob/f49504674362639fa312d9b5bd8ecf8d8f93ddbd/output/cloud/expv2/output.go#L116) the value has to be configured manually and explicitly.

So manual adjustments to `GOMAXPROCS` and `GOMEMLIMIT` are required to tune K6 to work best with the available resources of the system (hardware, VM, container).

I use k6s on Kubernetes and via the [k6-operator](https://github.com/grafana/k6-operator) and the provided `TestRun` CustomResource does allow for [requests and limits](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#resource-requests-and-limits-of-pod-and-container) to be configured. While these settings are then applied to the PodSpec, they are not converted or made available to the corresponding golang settings `GOMAXPROCS` and `GOMEMLIMIT`.

While this capability could be added to the k6-operator (shout out to @yorugac), to me it makes more sense to enable k6 itself to auto-tune itself to the available resources as it can be run without the k6 operator or non-containerized and will still benefit from these auto-tuning capabilities. There is a similar suggestion for the Grafana Helm chart, also discussing the use of auto-tuning for the options: https://github.com/grafana/helm-charts/pull/3138 (@jnoordsij, @jkroepke)

Therefore I'd like to propose adding automatic config of the Golang runtime via well established golang libs, which are actually
commonly used for other Grafana tools:

**GOMAXPROCS**
* https://github.com/uber-go/automaxprocs
* https://github.com/search?q=org%3Agrafana+automaxprocs+&type=code

**GOMEMLIMIT**
* https://github.com/KimMachineGun/automemlimit
* https://github.com/search?q=org%3Agrafana+automemlimit&type=code

There are many examples / MRs for the proposed mechanisms / libraries for other popular tools ....
* GOMAXPROCS for Prometheus https://github.com/prometheus/prometheus/pull/10498
* GOMEMLIMIT for Prometheus https://github.com/prometheus/prometheus/commit/2c0f9d1eee1ee68d830b4c2c070b3bf2b111c83e
* GOMAXPROCS for Kubernetes https://github.com/kubernetes/kubernetes/issues/116997
* https://gitlab.com/gitlab-org/gitlab-runner/-/merge_requests/4142
* ...

**Comment by olegbespalov** [comment]
@grafana/k6-core, what do you think?

**Comment by joanlopez** [comment]
> [@grafana/k6-core](https://github.com/orgs/grafana/teams/k6-core), what do you think?

I don't have specific experience with these libraries, but it definitely sounds as something that may worth looking at and trying out.
Although it isn't probably something with really high priority, tbh.

**Comment by SuperQ** [comment]
We use both (`uber-go/automaxprocs` and `KimMachineGun/automemlimit`) in Prometheus. They work very well to control resources.

**Comment by codebien** [comment]
Note, once Go 1.25 is released, this issue will become less relevant, because the default value for [GOMAXPROCS will then consider](https://pkg.go.dev/runtime@master#hdr-Default-GOMAXPROCS) the cgroup CPU quota.

**Comment by frittentheke** [comment]
> Note, once Go 1.25 is released, this issue will become less relevant, because the default value for [GOMAXPROCS will then consider](https://pkg.go.dev/runtime@master#hdr-Default-GOMAXPROCS) the cgroup CPU quota.

True, but this does not cover `GOMEMLIMIT` (discussed in https://github.com/golang/go/issues/75164 though).
Until then I believe https://github.com/KimMachineGun/automemlimit would be a nice and simple addition.
